backend/store/etcd/watcher.go

```diff
@@ -87,35 +86,30 @@ func (w *Watcher) start(ctx context.Context) {
 		defer close(w.resultChan)
 		_ = limiter.Wait(ctx)
 		for ctx.Err() == nil {
-			for watchResponse := range watcherChan {
+			select {
+			case watchResponse, closed := <-watcherChan:
 				if err := watchResponse.Err(); err != nil {
-					if ctx.Err() != nil {
-						// Our context was canceled, return without error,
-						// since the consumer is probably shutting down.
-						return
-					}
 					logger.WithError(err).Info("error from watch response")
 					w.resultChan <- store.WatchEvent{
 						Type: store.WatchError,
 						Err:  err,
 					}
-					if watchResponse.Canceled {
-						// Reinstate the watcher and break to the outer loop
+
+					if watchResponse.Canceled || closed {
```

Comment:
My reading of this is that this will result in the watcher being created for every watch event.

Reply:
This branch is inside the `if err := watchResponse.Err(); err != nil { }` branch; so the watcher is reinstated only if that `err != nil`.

Phrased another way, and assuming I rename that `closed` to `ok`, it means: if _any_ error was reported (`watchResponse.Err() != nil`) AND if the watch failed (`watchResponse.Canceled == true` according to etcd docs) OR if the channel is still open (`ok == true`) THEN just recreate the watcher.

Since the watcher can fail to close its channel or set `Canceled` to true, that's the only thing I tried that ended up working to not spin indefinitely...

I have a hard time explaining it even after spending that much time on it. Not a good sign...

Reply:
I see! This is indeed unintuitive. Thanks.